fairscale/optim/oss.py

```diff
@@ -146,9 +146,22 @@ def state_dict(self) -> Dict[str, Any]:
             len(self._all_states) > 0
         ), "The optimizer state is not materialized, please call consolidate_state_dict on every replica beforehand"
 
+        # Flatten the state_dict, save the partition
+        # The partition indexes the flat param_groups list,
+        # and logs the rank <> shard correspondence
+        partition: Dict[int, Tuple[int, int]] = {}
+        param_groups: List[Dict[Any, Any]] = []
+
+        i_global = 0
+        for i, s in enumerate(self._all_states):
+            param_groups.extend(s["param_groups"])
+            partition[i] = (i_global, i_global + len(s["param_groups"]))
+            i_global = partition[i][1]
```

Comment:
Might be simpler as:
```suggestion
            end = i_global + len(s["param_groups"])
            partition[i] = (i_global, end)
            i_global = end
```

Reply:
sure, just three lines to increment a counter, that's why I was trying to keep this compact. I miss the `index++` notation in C++ at times to be honest

Reply:
readability > compactness

Reply:
taste and colours, I was just trying to give the initial reasoning. Completely fine by me with a longer version